File: crates/taflow-core/src/stream/supertrend_test.rs
use super::Supertrend;

#[test]
fn supertrend_warmup_and_reset_are_deterministic() {
    let high: Vec<f64> = (0..96).map(|i| 100.0 + (i as f64 * 0.1).sin()).collect();
    let low: Vec<f64> = high.iter().map(|v| v - 2.0).collect();
    let close: Vec<f64> = high.iter().map(|v| v - 1.0).collect();
    let mut state = Supertrend::new(7, 3.0).unwrap();
    let first: Vec<f64> = high
        .iter()
        .zip(&low)
        .zip(&close)
        .map(|((&h, &l), &c)| state.append(h, l, c).map_or(f64::NAN, |v| v.trend))
        .collect();
    assert!(first[..6].iter().all(|v| v.is_nan()));
    assert!(first[6..].iter().all(|v| v.is_finite()));
    state.reset();
    let second: Vec<f64> = high
        .iter()
        .zip(&low)
        .zip(&close)
        .map(|((&h, &l), &c)| state.append(h, l, c).map_or(f64::NAN, |v| v.trend))
        .collect();
    assert_eq!(
        first.iter().map(|v| v.to_bits()).collect::<Vec<_>>(),
        second.iter().map(|v| v.to_bits()).collect::<Vec<_>>()
    );
}

File: tests/supertrend_test.py
import numpy as np

from taflow import Supertrend


def test_supertrend_lifecycle() -> None:
    high = 100.0 + np.sin(np.arange(96.0) / 10.0)
    low = high - 2.0
    close = high - 1.0
    state = Supertrend().extend(high, low, close)
    first = state.compute()
    assert np.isnan(first[0][:6]).all()
    state.reset().extend(high, low, close)
    for got, expected in zip(state.compute(), first):
        np.testing.assert_array_equal(got, expected)
    assert len(state) == len(close)

